### PR TITLE
- [FP-2787](https://movai.atlassian.net/browse/FP-2787): IDE - Topics - Can't open multiple Topics tabs

### PR DESCRIPTION
**Pull request by quirinpa**
- [FP-2787](https://movai.atlassian.net/browse/FP-2787): IDE - Topics - Can't open multiple Topics tabs
  - Notes:
    - See [this comment](https://movai.atlassian.net/browse/FP-2787?focusedCommentId=32319) if you have any doubts
    - Needs: https://github.com/MOV-AI/frontend-npm-ide/pull/594 (with updated lib-ide)

[FP-2787]: https://movai.atlassian.net/browse/FP-2787?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by diasdm — DISMISSED** [review]

**Review by diasdm — APPROVED** [review]

**Review by diasdm — CHANGES_REQUESTED** [review]
Take a look at the Sonar issues pls

File: src/utils/generalFunctions.js [hunk 1]
```diff
@@ -73,8 +74,21 @@ export function aboutPopup(call, classes = {}) {
...
+  const tabsMap = workspace.getTabs();
+  const toolIds = tabsMap.get("toolIds") ?? {};
+  const toolIdData = toolIds[toolName] ?? { last: 0, free: [] };
+  const id = toolIdData.free.shift() || toolIdData.last++;
+  const tabData = { ...getToolTabData({ scope: toolName, id: toolName + "-" + id }, props) };
+  if (tabData.multiple) {
+    tabData.name += "-" + id;
+    tabData.title += "-" + id;
+  } else
+    tabData.id = toolName;
```
Comment (diasdm) [inline src/utils/generalFunctions.js:87]: Can this be moved to getToolTabData?

**Review by diasdm — CHANGES_REQUESTED** [review]
Pls compute the toolId when loading the page

File: src/plugins/views/Tabs/hooks/useTabLayout.js [hunk 2]
```diff
@@ -655,9 +667,10 @@ const useTabLayout = (props, dockRef) => {
         isDirty,
         tabProps
       });
-      const tabData = { id, scope, name, tabTitle, extension };
+      const tabData = { id, rid, scope, name, tabTitle, extension };
       return {
         id: id,
+        rid: rid,
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/plugins/views/Tabs/hooks/useTabLayout.js:673]: realId, yes, but I think it isn't needed.

File: src/tools/index.js [hunk 3]
```diff
@@ -9,12 +9,12 @@ export const addTool = (name, tool) => {
 };
 
 export const getToolTabData = (tab, props = {}) => {
-  const { id, name } = tab;
-  const notInstalledTab = getTabData(id, name);
-  const data = id in APP_TOOLS ? APP_TOOLS[id].tabData : notInstalledTab;
+  const { id, name, scope } = tab;
+  const notInstalledTab = getTabData(scope, name);
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/tools/index.js]: Done

File: src/utils/generalFunctions.js (hunk 1 above)
Comment (quirinpa) [reply src/utils/generalFunctions.js:87]: It can but I don't think it should

File: src/plugins/views/Tabs/hooks/useTabLayout.js [hunk 4]
```diff
@@ -356,6 +356,16 @@ const useTabLayout = (props, dockRef) => {
 
         // Remove tab and apply new layout
         tabsById.current.delete(tabId);
+        const nid = new Number(tabId.substring(tabId.lastIndexOf("-") + 1));
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/plugins/views/Tabs/hooks/useTabLayout.js]: Done

File: src/utils/generalFunctions.js [hunk 5]
```diff
@@ -73,8 +74,21 @@ export function aboutPopup(call, classes = {}) {
...
  */
 export function openTool(call, toolName = getHomeTab().id, props = {}) {
-  const tabData = getToolTabData({ id: toolName }, props);
+  const workspace = new Workspace();
+  const tabsMap = workspace.getTabs();
+  const toolIds = tabsMap.get("toolIds") ?? {};
+  const toolIdData = toolIds[toolName] ?? { last: 0, free: [] };
+  const id = toolIdData.free.shift() || toolIdData.last++;
+  const tabData = { ...getToolTabData({ scope: toolName, id: toolName + "-" + id }, props) };
+  if (tabData.multiple) {
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/utils/generalFunctions.js:83]: We've talked about this. I think in the context we could agree it refers to tabs

**Comment by quirinpa** [comment]
A comment is missing. David asked to compute the toolIds object instead of loading from localStorage

**Review by manuelnogueiramov — APPROVED** [review]
So I did this review BEFORE you went on vacation, and I kinda forgot to click submit review :facepalm: 
I'm ok with this one as is :) 
Good work :+1:

File: src/utils/generalFunctions.js [hunk 6]
```diff
@@ -67,14 +68,76 @@ export function aboutPopup(call, classes = {}) {
...
+export function freeToolId(tabId) {
+  const [alpha, number] = breakToolId(tabId);
+
+  if (!number)
+    return;
+
+  const toolIdData = toolIds[alpha];
+  toolIdData.free.unshift(number);
+}
+
```
Comment (manuelnogueiramov) [inline src/utils/generalFunctions.js:123]: Actually coming to review this again, I'm trying to understand this code, and looks so much more, less performant and more complex than what I did [here](https://github.com/MOV-AI/frontend-npm-lib-ide/pull/334/files) in the `src/utils/Utils.js` to achieve the same thing. Can you confirm @EdwardAngeles @quirinpa ?

Comment (quirinpa) [reply src/utils/generalFunctions.js:123]: That only runs when the page loads.
It is to compute the things instead of using localStorage.
After the initial time it runs, it's pretty fast.

File: src/utils/generalFunctions.js [hunk 7]
```diff
@@ -67,14 +68,93 @@ export function aboutPopup(call, classes = {}) {
...
+  const res = {};
+
+  /* get a map of this format:
+   * {
+   *   Topics: { last: 3, busy: [2] },
+   *   VarsDebug: { last 2, busy: [1] },
+   * }
+   */
+  for (const [tabId] of tabsMap) {
+    const [alpha, number] = breakToolId(tabId);
```
Comment (manuelnogueiramov) [inline src/utils/generalFunctions.js:97]:
```suggestion
    const [toolName, tabIncrement] = breakToolId(tabId);
```

Comment (manuelnogueiramov) [reply src/utils/generalFunctions.js:97]: Also, if you get rid of the `breakToolId` function, you can use something like:

```
const {scope, tabIncrement} = tabData;
```

File: src/utils/generalFunctions.js [hunk 8]
```diff
@@ -67,14 +68,93 @@ export function aboutPopup(call, classes = {}) {
...
+    delete specific.busy;
+    specific.last++;
+  }
+
+  return res;
+}
+
+const workspace = new Workspace();
+const tabsMap = workspace.getTabs();
+const toolIds = computeIds(tabsMap);
```
Comment (manuelnogueiramov) [inline src/utils/generalFunctions.js:128]: I hate that we are just floating around in the middle of this `generalFunctions` file. I think you might be able to come up with some better way to do this.
Also, maybe all this "tool tab" logic should be in its own file, since none of these are actual "general functions" but very specific ones.

**Review by manuelnogueiramov — CHANGES_REQUESTED** [review]
Made a couple of suggestions that may improve the code. Take a look and let me know if you have any questions or would like to do a small pair session. 

tl;dr;
Overall, I like the logic of your thinking. I just think you made it harder to understand with generic, non-related naming in most vars and I don't think all this code belongs in this file.

File: src/utils/generalFunctions.js [hunk 9]
```diff
@@ -67,14 +68,93 @@ export function aboutPopup(call, classes = {}) {
...
+  return [alpha, number];
+}
+
+/* get the structure from the tabsMap that will allow us
+ * to efficently organize tool ids.
+ */
+function computeIds(tabsMap) {
+  const res = {};
+
+  /* get a map of this format:
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/utils/generalFunctions.js:90]: We are computing. But also creating.
I don't see this as a strong reason to hold the PR

File: src/utils/generalFunctions.js [hunk 10]
```diff
@@ -67,14 +68,93 @@ export function aboutPopup(call, classes = {}) {
...
+   *   Topics: { last: 3, busy: [2] },
+   *   VarsDebug: { last 2, busy: [1] },
+   * }
+   */
+  for (const [tabId] of tabsMap) {
+    const [alpha, number] = breakToolId(tabId);
+    const index = tabId.search(/\d/);
+    if (index < 0)
+      continue;
+    const specific = res[alpha] ?? { last: 0, busy: {}, free: [] };
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/utils/generalFunctions.js:101]: The resulting object is not a tool tab. But indeed something specific to that tool.

File: src/utils/generalFunctions.js [hunk 11]
```diff
@@ -67,14 +68,93 @@ export function aboutPopup(call, classes = {}) {
...
+  const number = new Number(tabId.substring(index));
+
+  return [alpha, number];
+}
+
+/* get the structure from the tabsMap that will allow us
+ * to efficently organize tool ids.
+ */
+function computeIds(tabsMap) {
+  const res = {};
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/utils/generalFunctions.js:88]: res comes from "result". I don't think "toolTabIncrementMap is a good name.

File: src/utils/generalFunctions.js (hunk 7 above)
Comment (quirinpa) [reply src/utils/generalFunctions.js:97]: I don't think there's a strong reason to modify what we're storing there.

File: src/utils/generalFunctions.js (hunk 8 above)
Comment (quirinpa) [reply src/utils/generalFunctions.js:128]: This code is only used here. We already have too many files.

**Comment by diasdm** [comment]
Replaced by https://github.com/MOV-AI/frontend-npm-lib-ide/pull/334
